### PR TITLE
Support multi-statement execution for PostgreSQL

### LINKED ISSUE #284: Support creating multiple indexes concurrently in a single migration postgresql
**Describe the Bug**
I'd like to add multiple indexes to my postgresql schema in a single migration file.

**Steps to Reproduce**
1. My migrations look like '...'

```
CREATE INDEX CONCURRENTLY idx_my_table_col1 ON my_table(col1);
CREATE INDEX CONCURRENTLY idx_my_table_col2 ON my_table(col2);
```

2. I ran migrate with the following options '....'

```
migrate -source file:migrations -database "postgres://localhost:5432/my-database?sslmode=disable" up
```

3. See error
```
 (details: pq: CREATE INDEX CONCURRENTLY cannot run inside a transaction block)
```

**Expected Behavior**
I expected the migration to succeed and add both indexes. This migration does succeed if I only try to create one index at a time and have two sequential migration files with the following contents:

``````
CREATE INDEX CONCURRENTLY idx_my_table_col1 ON my_table(col1);
```
```
CREATE INDEX CONCURRENTLY idx_my_table_col2 ON my_table(col2);
```

**Migrate Version**
v4.6.2

**Loaded Source Drivers**
Source drivers: github, github-ee, gitlab, go-bindata, godoc-vfs, gcs, file, s3

**Loaded Database Drivers**
Database drivers: cockroachdb, crdb-postgres, postgres, redshift, postgresql, spanner, sqlserver, cassandra, clickhouse, cockroach, mongodb, mysql, stub


**Go Version**
go version go1.13 darwin/amd64

### LINKED ISSUE #492: PostgreSQL COPY ... FROM STDIN throws a syntax error
**Describe the Bug**
I'm trying to apply a migration with `COPY ... FROM STDIN` command inside and get a syntax error. But actually, there is no problem with SQL syntax, because I can apply the same migration manually or with other tools like `FlyWay`.

**Steps to Reproduce**
An environment described below present in this archive: [test.tar.gz](https://github.com/golang-migrate/migrate/files/5747030/test.tar.gz).

1. Use this docker-compose manifest:
    ```
    ---
    version: '3.3' 

    services:
      db:
        container_name: db
        hostname: db
        image: postgres:12
        ports: 
          - 5432:5432
        environment:
          - POSTGRES_DB=test
          - POSTGRES_PASSWORD=postgres
        networks:
          - migrate

      migrate:
        container_name: migrate
        hostname: migrate
        image: migrate/migrate:v4.14.1
        entrypoint: ''
        command: ['sh', '-c', 'sleep 5 && migrate -path=/sql -database=postgres://postgres:postgres@db/test?sslmode=disable up']
        volumes:
          - ./sql:/sql
        depends_on:
          - db
        networks:
          - migrate

    networks:
      migrate:
        driver: bridge

    ```
  
2. Put these migration files into `sql` folder near the docker-compose manifest:
    - 1_schema.up.sql:
      ```
      create table sample (
          id int primary key,
          text varchar(64) not null unique
      ); 
      ```
    - 2_data.up.sql:
      ```
      COPY sample (id, text) FROM stdin;
      10  Some text 1
      20  Some text 2
      30  Some text 3
      \.

      ```

3. Run the `docker-compose up -d` command and see logs from the `migrate` container: `docker-compose logs migrate` and you will see the next error:
    ```
    Attaching to migrate
    migrate    | 1/u schema (25.207327ms)
    migrate    | error: migration failed: syntax error at or near "10" (column 1) in line 2: COPY sample (id, text) FROM stdin;
    migrate    | 10 Some text 1
    migrate    | 20 Some text 2
    migrate    | 30 Some text 3
    migrate    | \.
    migrate    |  (details: pq: syntax error at or near "10")
    ```

**Expected Behavior**
Migration file with `COPY ... FROM STDIN` should be applied successfully.

**Migrate Version**
v4.14.1

**Loaded Source Drivers**
- go-bindata
- godoc-vfs 
- gcs
- file
- bitbucket
- github
- github-ee
- gitlab, s3

**Loaded Database Drivers**
- mongodb
- mysql
- sqlserver
- stub
- clickhouse
- crdb-postgres
- firebird
- postgresql
- redshift
- cassandra
- postgres
- cockroach
- cockroachdb
- mongodb+srv
- neo4j
- spanner

**Go Version**
There is no Go installed inside the official Docker image used.

**Stacktrace**
There are no stacktraces.

**Additional context**
There is no additional context.

Is it a bug or I do something wrong in `migrate` usage?

### PR DESCRIPTION
**Pull request by AnatolyRugalev**
This PR adds support for multi-statement SQL execution for PostgreSQL driver.

In PostgreSQL running multiple SQL statements in one `Exec` executes them inside a transaction. Sometimes this behavior is not desirable because some statements can be only run outside of transaction (e.g. `CREATE INDEX CONCURRENTLY`). If you want to use `CREATE INDEX CONCURRENTLY` without activating multi-statement mode you have to put such statements in a separate migration files.

* [x] Provide some tests

Resolves #284, Resolves #492

### CODE REVIEW

**Comment by AnatolyRugalev** [comment]
Hey @dhui!

Can you please take a look? This is mostly a carbon copy from clickhouse multi-statement implementation. The init code, and constants duplicate each other across the drivers, so probably not a bad idea to extract this into a common place (or even introduce new `MultiStatementDriver` interface).

Also I'm not sure how the multi-statement parser will split something like this:

```sql
INSERT INTO foo (bar)  VALUES (';');
```

Shall we improve it in a separate issue and PR?

File: database/postgres/postgres.go
Comment (dhui) [inline database/postgres/postgres.go:143]: > The init code, and constants duplicate each other across the drivers, so probably not a bad idea to extract this into a common place (or even introduce new `MultiStatementDriver` interface).

Is this what you meant by init code?
What were you thinking in terms of a `MultiStatementDriver` interface?

File: database/postgres/postgres.go
Comment (dhui) [inline database/postgres/postgres.go:244]: Run the original statement with all of the whitespace and limit `strings.TrimSpace()` to only be used in the empty query check.

File: database/postgres/postgres_test.go
Comment (dhui) [inline database/postgres/postgres_test.go:165]: Instead of modifying the returned connection string, do all of the url/addr building in `pgConnectionString()`

**Review by dhui — CHANGES_REQUESTED** [review]
Thanks for the PR!

> Also I'm not sure how the multi-statement parser will split something like this:
> `INSERT INTO foo (bar)  VALUES (';');`
> Shall we improve it in a separate issue and PR?

Agreed that this is an issue and is out of scope for this PR.
How would you work around this issue w/o writing a mini SQL parser? There are other cases with invalid SQL and various escape sequences that will break the simple and naive `;` splitting that `multistmt` does.
`migrate` [intentionally](https://github.com/golang-migrate/migrate/blob/master/MIGRATIONS.md#migration-content-format) avoids parsing migrations with `multistmt` being the exception due to the need by various DB drivers and the simplicity of the approach.

FYI, unfortunately, merges are blocked until CI is fixed. See: https://github.com/golang-migrate/migrate/issues/252#issuecomment-747206381
However, if this PR looks good. I'll approve and will merge it once CI is fixed.

**Review by dhui — APPROVED** [review]
Thanks for addressing my feedback!

**Comment by andream16** [comment]
I was hoping that this PR would also address this error I'm getting:

`CREATE INDEX CONCURRENTLY cannot be executed from a function or multi-command string in line 0: SET statement_timeout = "24h";`

This is the migrations file looks like:

```
SET statement_timeout = "24h";
CREATE INDEX CONCURRENTLY IF NOT EXISTS my_index ON my_table(my_column);
SET statement_timeout = default;
```

Any suggestions?

**Comment by AnatolyRugalev** [comment]
@andream16 did you put `x-multi-statement=true` to the connection string?

**Comment by andream16** [comment]
@AnatolyRugalev thanks for the reply. 

I'm using `migrate.NewWithDatabaseInstance()` to get a new pointer to `Migrate`. 

I'm using `pq` to create the instance and it apparently it doesn't support such parameter in the dsn. The creation fails with: `pq: unrecognized configuration parameter "x-multi-statement"`

I guess I'll have to try using `migrate.New()` directly.

**Comment by AnatolyRugalev** [comment]
@andream16 if you have `sql.DB` ready to use with migrate you can probably do this:

```go
var db *sql.DB
... // I assume db gets initialized here

pgDriver := postgres.WithInstance(db), postgres.Config{
    MultiStatementEnabled: true
})
migrate.NewWithDatabaseInstance(sourceURL, dbName, pgDriver)
```

**Comment by andream16** [comment]
@AnatolyRugalev that works! Thank you 🎉 

For other folks having this problem - you can also achieve the same result using `migrate.New()`:

```
...

import (
    ...
    "github.com/golang-migrate/migrate/v4"
    _ "github.com/golang-migrate/migrate/v4/database/postgres"
    _ "github.com/golang-migrate/migrate/v4/source/file"
)

func NewMigrate(migrationsPath, dsn) (*migrate.Migrate) {
    m, err := migrate.New("file://"+migrationsPath, dsn)
    if err != nil {
        return nil, fmt.Errorf("could not create new migrate: %w", err)
    }
    
    return m, nil
}

...
_, _ = NewMigrate("/migrations", "postgres://user:password@hostname:port/db_name?sslmode=disable&x-multi-statement=true")
```

**Pull request merged**
